### DISCUSSION TITLE
Pass a value to event scripts

### DISCUSSION THREAD

**Discussion by neilenns** [MobiFlight/MobiFlight-Connector#420] [Ideas]
(From a discussion on Discord)

There are lots of switches in the FS2020 aircraft that work off a `_SET` event, e.g. `(>K:BEACON_LIGHTS_SET)`. This results in a ton of very simple scripts like the following:

```
G36_LIGHT_BEACON_ON#1 (>K:BEACON_LIGHTS_SET)
```

It would be nice if there was a way to specify a value in the MF UI that gets passed to the script and inserted as a "local" variable. Then the script could be something like this:

```
G36_LIGHT_BEACON_SET#(L:MobiFlight-UI-Value) (>K:BEACON_LIGHTS_SET)
```

The switch configuration in MF would use the same script for both Press and Release states, with a different value set in the UI for the variable (1 for press, 0 for release).

This is similar to how AxisAndOhs handles this and cuts the number of scripts required in half.

**Comment by neilenns** [comment]
A variation on this idea that would get rid of the need for custom scripts entirely is to build the script on the fly if it's a basic K value and something is set in the UI. MF could construct `0 (>K:BEACON_LIGHTS_SET)` on the fly if the script is nothing more than a K set clause and a value in the UI.

**Reply by neilenns** [reply]
While this does skip the need for custom scripts, I have to admit I'm partial to the UI + basic custom script. The huge value in the custom scripts we're figuring out is they are clearly marked for the aircraft they support and it makes it really easy to know you've got the right event.

If I'm new to MobiFlight and SimConnect it's *much* easier for me to know that `G36_LIGHT_BEACON_SET` is the thing I need for the G36 than wondering if `(>K:BEACON_LIGHTS_SET)` is the right thing to use.

Heck, even though it's repetitive the one time we make the event, `G36_LIGHT_BEACON_ON` and `G36_LIGHT_BEACON_OFF` is even more clear for new users than _SET. Hmm...

**Reply by tigert** [reply]
It is clear, though if you have _ON and _OFF events, you *must* define both ON and OFF states on your switch. With a parameter (like it works with EventID and FSUIPC) the switch state itself is sent as a parameter by default and it mostly works right.

**Reply by neilenns** [reply]
Sending the switch state by default is a nice variation on a solution as well!

**Reply by MobiFlight-Admin** [reply]
On my last stream I actually got this all to work. I think it makes sense to add it. We can make it really flexible, but only if a user really needs it. The presets are better for most users.

**Comment by MobiFlight-Admin** [comment]
For presets for me the preferred way is to provide concise and easy to understand out-of-the-box working event presets. A user should not need to provide a number dynamically, because that can already be overwhelming in the beginning.

There is one use-case where the parameter has to be provided dynamically and that is for the potentiometer.
For this use-case however we can provide the MSFS2020 Custom Input Action (https://github.com/Mobiflight/MobiFlight-Connector/wiki/Input-Actions#msfs2020---custom-input).

For FSUIPC - EventIDs we have solved this by providing an extra "param" field. This would also be OK if we were only talking about K and H events. But our scripts are already very complicated, maybe with the need to replace not just one variable, but also ideally being able to use other placeholders, etc.

For now, I will let the Events the way they are... self-contained, no user input needed.
For user input based stuff someone can use MSFS 2020 Custom Input Actions.
Eventually both approaches should merge into one, where the Custom Input Actions simply come with the presets for the events and then you can edit them on the fly if you would want to.

**Comment by DocMoebiuz** [comment]
this is now totally possible as of 9.2+
